### DISCUSSION TITLE
Special / Inherited Members in anonymous type

### DISCUSSION THREAD

**Discussion by bernd5** [dotnet/csharplang#4417] [Q&A]
I found out, that roslyn does allow that anonymous types have properties of inherited member names [like](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAgrgOwD4AEBMAjALABQZAbhAgAQAm+NAvDZnAO40DeNAKgPYBlGAgCWmAObMa+JDQDicGAAloACwDC/OnGkBmOQFEAjuggAbKNIAsc3PcYsArHNgAddAAZPANkIxRAGNpAHY5RRheAE8AB10WAA4aAF8Abkpqej1pNk4eAWExSWlZBSVVKE1teJoDGhMzSxs7B2kXGncvX38g0PClaLjpJLSyDNo1QmkGADoIiqqdAAoASnTSKgnHenw58vUtZbXxmjVsljo9PZUD6tX1slxCAE4l2YaLKDe9FePSJ9eDGYFx+D1IogAZjQlpNgad8CsyFwyDRUTQAUsAEQATTgUExf2SZEh0NhTBYZ0RpGRpDR6JeWIAIo4WIy9AT1kTyBtMrBaBddgUROIJKsaABqHYzIVFKSS2YRQa6eW7RWxOD3YlQt67D6WHW/JEotEYnF4jlkLmPBl8tZAA):

```CSharp
var d2 = new { ToString = 2, GetHashCode = 3, Equals = 4, GetType = 5 };
```

**Is this required to work?**
If you paste this code for example in a VisualStudio installation with ReSharper installed you get red squiggles:
![grafik](https://user-images.githubusercontent.com/26689688/107576702-cae18600-6bf1-11eb-8eaa-102743217c5a.png)

If you apply the Resharper-Reforctoring "Anonymous to Named type", you get the following result:

```CSharp
public class SomeAnonymousType
{
    public int ToString { get; set; }
    public int GetHashCode { get; set; }
    public int Equals { get; set; }

    public override string ToString()
    {
        return $"{{ ToString = {ToString}, GetHashCode = {GetHashCode}, Equals = {Equals} }}";
    }

    public override bool Equals(object value)
    {
        var type = value as SomeAnonymousType;
        return (type != null) && EqualityComparer<int>.Default.Equals((int) type.ToString, 
            (int) ToString) && EqualityComparer<int>.Default.Equals(type.GetHashCode, GetHashCode) && 
               EqualityComparer<int>.Default.Equals((int) type.Equals, (int) Equals);
    }

    public override int GetHashCode()
    {
        int num = 0x7a2f0b42;
        num = (-1521134295 * num) + EqualityComparer<int>.Default.GetHashCode(ToString);
        num = (-1521134295 * num) + EqualityComparer<int>.Default.GetHashCode(GetHashCode);
        return (-1521134295 * num) + EqualityComparer<int>.Default.GetHashCode(Equals);
    }
}
```

Unfortunately this code does not compile. Mainly because of 

> "CS0102: The type 'SomeAnonymousType' already contains a definition for 'ToString'"

Why does the overrides conflicts with the properties?
If you take a look into the generated IL Code produced by Roslyn (e.g. with ILSpy) the code looks very similar.

On the other hand if you split the anonymous type in two classes it is possible to define something similar (simplified - just implementing `Equals`):

```CSharp
class AnonBase
{
    public override bool Equals(object obj)
    {
        if (obj is AnonExplicit d)
        {
            return Equals((AnonExplicit) this, d);
        }
        return false;
    }
    
    public static bool Equals(AnonExplicit me, AnonExplicit obj)
    {
        if (!Equals(me.Equals,obj.Equals))
        {
            return false;    
        }
        return true;
    }
}


class AnonExplicit : AnonBase
{
    public int Equals { get; set; }
}

class App{
    
    static void Main(){
        var anon = new { Equals = 42 };
        Console.WriteLine(anon.Equals(anon));
        Console.WriteLine(anon.Equals);
    }
}
```

**Comment by HaloFour** [comment]
IL allows you to have a property and a method that has the same name.  C# does not.  The anonymous class feature of the compiler is taking advantage of features that you're not permitted to use directly.

While it's probably a bit pathological to name the elements of an anonymous type after the methods of `System.Object` or that are generated by the compiler it's probably quite a bit too late to change that since it would be a breaking change.

**Reply by bernd5** [reply]
In records for example those names can't be used - I'm more interested if a compiler is required to support those names to be language / spec conformant or if it is just a really pathological case.

**Reply by CyrusNajmabadi** [reply]
> I'm more interested if a compiler is required to support those names t

Yes.  As these names are not disallowed, the compiler is required to support them.
